File: rateLimiter/go-service/internal/limiter/sliding_window_counter.go
package limiter

import (
	"context"
	"fmt"
	"math"
	"time"
)

type SlidingWindowCounterLimiter struct {
	Store         CounterStore
	Limit         int
	WindowSeconds float64
}

func (l *SlidingWindowCounterLimiter) Allow(ctx context.Context, key string) (Result, error) {
	now := float64(time.Now().UnixNano()) / 1e9
	windowIndex := int64(math.Floor(now / l.WindowSeconds))
	windowStart := float64(windowIndex) * l.WindowSeconds
	weight := 1 - (now-windowStart)/l.WindowSeconds

	currentKey := fmt.Sprintf("%s:%d", key, windowIndex)
	previousKey := fmt.Sprintf("%s:%d", key, windowIndex-1)

	// Read-both-and-conditionally-increment as ONE atomic store call - see
	// stores.go's CounterStore doc comment for why a separate
	// get-then-increment here would reintroduce the exact race this
	// project exists to prevent.
	allowed, estimate, err := l.Store.IncrementWeighted(
		ctx, currentKey, previousKey, weight, l.Limit, l.WindowSeconds*2,
	)
	if err != nil {
		return Result{}, err
	}

	remaining := max(0, int(float64(l.Limit)-estimate))
	windowEnd := windowStart + l.WindowSeconds

	result := Result{Allowed: allowed, Limit: l.Limit, Remaining: remaining, ResetAt: windowEnd}
	if !allowed {
		retryAfter := windowEnd - now
		result.RetryAfter = &retryAfter
	}
	return result, nil
}

File: rateLimiter/fastapi-service/app/limiter/sliding_window_counter.py
import time
from typing import Optional

from app.limiter.base import RateLimitResult
from app.storage.base import CounterStore


class SlidingWindowCounterLimiter:
    def __init__(self, store: CounterStore, limit: int, window_seconds: float) -> None:
        self._store = store
        self._limit = limit
        self._window_seconds = window_seconds

    async def allow(self, key: str) -> RateLimitResult:
        now = time.time()
        window_index = int(now // self._window_seconds)
        window_start = window_index * self._window_seconds
        weight = 1 - ((now - window_start) / self._window_seconds)

        current_key = f"{key}:{window_index}"
        previous_key = f"{key}:{window_index - 1}"

        # Read-both-and-conditionally-increment as ONE atomic store call -
        # see storage/base.py's CounterStore docstring for why a separate
        # get-then-increment here would reintroduce the exact race this
        # project exists to prevent.
        allowed, estimate = await self._store.increment_weighted(
            current_key, previous_key, weight, self._limit, self._window_seconds * 2
        )

        remaining = max(0, int(self._limit - estimate))
        window_end = window_start + self._window_seconds
        retry_after: Optional[float] = (window_end - now) if not allowed else None

        return RateLimitResult(
            allowed=allowed,
            limit=self._limit,
            remaining=remaining,
            reset_at=window_end,
            retry_after=retry_after,
        )
